include/nil/actor/zk/commitments/detail/polynomial/proof_of_work.hpp

```diff
@@ -83,6 +87,96 @@ namespace nil {
                         return ((result & mask) == 0);
                     }
                 };
+
+                // Note that the interface here is slightly different from the one above:
+                // amount of bits for grinding instead of the mask.
+                // This was done because the actual mask is applied to the high bits instead of the low bits
+                // which makes manually setting the mask error-prone.
+                template<typename TranscriptHashType, typename FieldType, std::uint8_t GrindingBits=16>
+                class field_proof_of_work {
+                public:
+                    using transcript_hash_type = TranscriptHashType;
+                    using transcript_type = transcript::fiat_shamir_heuristic_sequential<transcript_hash_type>;
+                    using value_type = typename FieldType::value_type;
+                    using integral_type = typename FieldType::integral_type;
+
+                    constexpr static const integral_type mask =
+                        (GrindingBits > 0 ?
+                            ((integral_type(2) << GrindingBits - 1) - 1) << (FieldType::modulus_bits - GrindingBits)
+                            : 0);
+
+                    static inline boost::property_tree::ptree get_params() {
+                        boost::property_tree::ptree params;
+                        params.put("mask", mask);
+                        return params;
+                    }
+
+                    static inline value_type generate(transcript_type &transcript, 
+                        nil::crypto3::random::algebraic_engine<FieldType> random_engine) {
+
+                        value_type pow_seed = random_engine();
+                        value_type init_seed = pow_seed;
+                        std::size_t per_core = 4*65536;
+                        std::vector<value_type> pow_values(smp::count*per_core);
```

Comment:
I don't get why we need this array "pow_values".

Reply:
updated, removed the array.